src/lib/facit.ts

```diff
@@ -0,0 +1,77 @@
+import csv from 'csv-parser'
+import fs from 'fs'
+import path from 'path'
+import { CompanyData } from '../models/companyEmissions'
+
+const parse = (str) =>
+  parseFloat(str.replaceAll(',', '').replace(/[^0-9.]/g, ''))
+
+// Scope 1     ,Scope 2 (LB) ,Scope 2 (MB) ,Scope 1+2   ,Scope 3 (total)
+function mapFacitToCompanyData(row): CompanyData {
+  return {
+    companyName: row['Company'],
+    emissions: [
+      {
+        year: 2023,
+        scope1: {
+          emissions: parse(row['23_Scope 1']),
+        },
+        scope2: {
+          lb: parse(row['23_Scope 2 (LB)']),
+          mb: parse(row['23_Scope 2 (MB)']),
+          emissions: parse(row['23_Scope 1+2']),
+        },
+        scope3: {
+          emissions: parse(row['23_Scope 3 (total)']),
+          categories: {
+            'Category 1': parse(row['Cat 1']),
+            'Category 2': parse(row['Cat 1']),
+            'Category 3': parse(row['Cat 3']),
+            'Category 4': parse(row['Cat 4']),
+            'Category 5': parse(row['Cat 5']),
+            'Category 6': parse(row['Cat 6']),
+            'Category 7': parse(row['Cat 7']),
+            'Category 8': parse(row['Cat 8']),
+            'Category 9': parse(row['Cat 9']),
+            'Category 10': parse(row['Cat 10']),
+            'Category 11': parse(row['Cat 11']),
+            'Category 12': parse(row['Cat 12']),
+            'Category 13': parse(row['Cat 13']),
+            'Category 14': parse(row['Cat 14']),
+            'Category 15': parse(row['Cat 15']),
+          },
+        },
+      },
+    ],
+  }
+}
+
+export function findFacit(url: string): Promise<CompanyData> {
+  const all = []
+  return new Promise((resolve, reject) => {
+    try {
+      fs.createReadStream(path.join(process.cwd(), 'src/data/facit.csv'))
+        .pipe(
+          csv({
+            mapHeaders: ({ header }) => header.trim(),
+            mapValues: ({ value }) => value.trim(),
+          })
+        )
+        .on('data', (data) => all.push(data))
+        .on('end', () => {
+          const found = all.find((result) => result['URL 2023'] === url)
+          console.log('found', found)
+          resolve(mapFacitToCompanyData(found))
```

Comment:
Since we only want one row, it feels a bit wasteful to parse the entire csv file.

Since we get one row back at a time, I guess one solution could be to move the url check straight into the data callback:

```ts
.on('data', (data) => {
  if (data['URL 2023'] === url) {
    // IDEA: Could we abort the readable stream as soon as we find the data we want?
    // For that we might need to get a reference to it, but it must be doable right?

    resolve(mapFacitToCompanyData(found))
  }
})
```

With this solution, we wouldn't need the `all` array, or the `.on('end', () => {})` callback.

A similar idea is if we could read the csv file only once and keep the parsed data in memory? Or is it worth keeping the memory footprint as low as possible? In that case maybe just the solution above is sufficient.

In the future with many thousands of companies, the facit should probably anyway be moved to a DB of some kind to make it easier to query.

Reply:
This is a solution only for now- as you say- we will probably use a database for this later on. Parsing from memory takes milliseconds so I wouldnt say we need to worry about this right now.